src/stratigraphy/util/geometric_line_utilities.py

```diff
@@ -223,36 +229,106 @@ def merge_parallel_lines(lines: list[Line], tol: int = 8, angle_threshold: float
 
 
+def odr_regression(x: ArrayLike, y: ArrayLike) -> tuple[float, float]:
+    """Perform orthogonal distance regression on the given data.
+
+    Solves the line equation y = mx + c for the best fit values of m and c.
+    Different to linear regression, ODR takes into account the error in both x and y.
+    As such, it can handle horizontal and vertical lines.
+
+    Note: For vertical lines, the slope is nan.
+
+    Args:
+        x (ArrayLike): The x-coordinates of the data.
+        y (ArrayLike): The y-coordinates of the data.
+
+    Returns:
+        tuple[float, float]: (m, c), the best fit values for the line equation.
+    """
+
+    # Define a function to fit the data with.
+    def linear_func(p, x):
+        m, c = p
+        return m * x + c
+
+    linear_model = odr.Model(linear_func)
+    data = odr.RealData(x, y)
+    odr_model = odr.ODR(data, linear_model, beta0=[0.0, 1.0])  # kept the initial guess as default
+
+    out = odr_model.run()
+
+    return out.beta[0], out.beta[1]
+
+
 def _merge_lines(line1: Line, line2: Line) -> Line:
-    """Merge two lines.
+    """Merge two lines into one.
+
+    The algorithm performes odr regression on the points of the two lines to find the best fit line.
+    Then, it calculates the orthogonal projection of the four points onto the best fit line and takes the two points
+    that are the furthest apart. These two points are then used to create the merged line.
 
-    Note: This method is not robust for near vertical lines. We assume
-    an ordering in the x-coordinates of the lines.
+    Note: Strictly vertical lines cannot be handled. In that case the slope is nan. A warning is thrown
+    and the first line is returned.
 
     Args:
-        line1 (Line): The first line.
-        line2 (Line): The second line.
+        line1 (Line): First line to merge.
+        line2 (Line): Second line to merge.
 
     Returns:
         Line: The merged line.
     """
-    # determine start point
-    if line1.start.x > line2.start.x:
-        x_start = line2.start.x
-        y_start = line2.start.y
-    else:
-        x_start = line1.start.x
-        y_start = line1.start.y
+    x = np.array([line1.start.x, line1.end.x, line2.start.x, line2.end.x])
+    y = np.array([line1.start.y, line1.end.y, line2.start.y, line2.end.y])
+    slope, intercept = odr_regression(x, y)
+    if np.isnan(slope):
+        logger.warning("Merged line slope is nan. Returning line1 in absence of any better choice.")
+        return line1
+    projected_points = []
+    for point in [line1.start, line1.end, line2.start, line2.end]:
+        projection = _get_orthogonal_projection_to_line(point, slope, intercept)
+        projected_points.append(projection)
+    # Take the two points such that the distance between them is the largest.
+    # Since there are only 4 points involved, we can just calculate the distance between all points.
+    point_combinations = list(combinations(projected_points, 2))
```

Comment:
I don't really like the quadratic complexity here, though for us it's probably not a problem because we only ever call this with four points.
I guess a more efficient method would be: take the points with min and max value for the x coordinates, unless the points all have equal x coordinates, in that case take the points with the min and max values for the y coordinates.

Reply:
Since the number of points is always four, I deemed it constant in time. And it executes fast. I don't see any use case where we feed more than 4 points to this function.